File: src/kinetic/consumer.clj
(ns kinetic.consumer
  (:require [clojure.tools.logging :as log]
            [kinetic.tools :as t]
            [kinetic.tracker :as tracker]
            [kinetic.aws :as aws])
  (:import [java.util UUID]
           [java.time Duration]
           [java.nio.charset StandardCharsets]
           [software.amazon.kinesis.common ConfigsBuilder
                                           InitialPositionInStream
                                           InitialPositionInStreamExtended
                                           StreamIdentifier
                                           StreamConfig]
           [software.amazon.kinesis.coordinator Scheduler]
           [software.amazon.kinesis.leases Lease]
           [software.amazon.kinesis.processor ShardRecordProcessor
                                              ShardRecordProcessorFactory
                                              FormerStreamsLeasesDeletionStrategy
                                              FormerStreamsLeasesDeletionStrategy$AutoDetectionAndDeferredDeletionStrategy
                                              StreamTracker
                                              SingleStreamTracker]
           [software.amazon.kinesis.retrieval KinesisClientRecord
                                              polling.PollingConfig]))

(defn make-config
  [{:keys [streams
           multi-stream?        ;; in case we need a multi stream tracking
           aws-account-number
           application-name
           kinesis-client
           dynamo-client
           cloud-watch-client
           record-processor-factory] :as opts}]
  (let [multi? (or multi-stream?
                   (> (count streams) 1))]
    {:config-builder (ConfigsBuilder. (tracker/make-stream-tracker opts)
                                      application-name
                                      kinesis-client
                                      dynamo-client
                                      cloud-watch-client
                                      (-> (UUID/randomUUID) str)
                                      record-processor-factory)
     :polling-config (if multi?
                       (PollingConfig. kinesis-client)
                       (PollingConfig. (-> streams first :name)
                                       kinesis-client))}))

(defn make-scheduler [{:keys [config-builder
                              polling-config]}]

  ;; TODO: add more params beyond the ConfigsBuilder defaults
  (Scheduler. (.checkpointConfig config-builder)
              (.coordinatorConfig config-builder)
              (.leaseManagementConfig config-builder)
              (.lifecycleConfig config-builder)
              (.metricsConfig config-builder)
              (.processorConfig config-builder)
              (-> config-builder
                  .retrievalConfig
                  (.retrievalSpecificConfig polling-config))))

(defn record-checkpoint [checkpointer]
  (let [last-sequence (.lastCheckpointValue checkpointer)
        current-sequence (.largestPermittedCheckpointValue checkpointer)]
    (log/infof "recording a checkpoint at sequence: %s (last recorded sequence was %s)"
               current-sequence last-sequence)
    (.checkpoint checkpointer)))

(defn ->record [^KinesisClientRecord record]
  {:sequence-number                (.sequenceNumber record)
   :approximate-arrival-timestamp  (.approximateArrivalTimestamp record)
   :data                           (.data record)
   :partitionKey                   (.partitionKey record)
   :encryption-type                (.encryptionType record)
   :sub-sequence-number            (.subSequenceNumber record)
   :explicit-hash-key              (.explicitHashKey record)
   :aggregated                     (.aggregated record)
   :schema                         (.schema record)})

(defn make-utf8-decoder []
  (let [decoder (-> (StandardCharsets/UTF_8)
                    .newDecoder)]
    (fn [{:keys [data]}]
      (->> data
           (.decode decoder)
           .toString))))

(defn shard-record-processor [consume
                              checkpoint-every-ms]
  (let [shard-id (atom "was not initialized with the shard id")]
    (reify ShardRecordProcessor

      (initialize
        [this init-with]
        "invoked by the Amazon Kinesis Client Library before data records are delivered to the ShardRecordProcessor instance
         (via 'processRecords').

         args: 'init-with' provides information related to initialization"
        (let [sid (.shardId init-with)
              sequence-number (.extendedSequenceNumber init-with)]
          (reset! shard-id sid)
          (log/infof "initializing shard %s at sequence %s"
                     @shard-id
                     sequence-number)))

      (processRecords
        [this batch]
        "process data records.
         the Amazon Kinesis Client Library will invoke this method to deliver data records to the application.
         upon fail over, the new instance will get records with sequence number > checkpoint position for each partition key.

         args: 'records' provides the records to be processed as well as information and capabilities related to them (eg checkpointing)."
        (consume (->> batch .records (mapv ->record)))
        ;; TODO: (when checkpoint-every-ms ...)
        (record-checkpoint (.checkpointer batch)))

      (leaseLost
        [this lost-lease]
        "called when the lease that tied to this record processor has been lost.
         once the lease has been lost the record processor can no longer checkpoint.

         args: 'lost-lease' access to functions and data related to the loss of the lease.
               currently this has no functionality.")

      (shardEnded
        [this completed-shard]
        "called when the shard that this record process is handling has been completed.
         once a shard has been completed no further records will ever arrive on that shard.

         when this is called the record processor must call 'record-checkpoint'
         otherwise an exception will be thrown and the all child shards of this shard will not make progress.

         args: 'completed-shard' provides access to a checkpointer method for completing processing of the shard."

        (log/infof "%s shard is completed. recording a checkpoint.." @shard-id)
        (try
          (-> completed-shard .checkpointer .checkpoint)
          (catch Exception e
            (log/errorf e "failed to checkpoint at shard end for %s" @shard-id)
            (throw e))))

      (shutdownRequested
        [this shutdown-request]
        "called when the Scheduler has been requested to shutdown.
         this is called while the record processor still holds the lease so checkpointing is possible.
         once this method has completed the lease for the record processor is released,
              and 'leaseLost' will be called at a later time.

         args: 'shutdown-request' provides access to a checkpointer (RecordProcessorCheckpointer)
               allowing a record processor to checkpoint before the shutdown is completed."

        (log/infof "scheduler for the shard %s is shutting down. recording a checkpoint.." @shard-id)
        (try
          (-> shutdown-request .checkpointer .checkpoint)
          (catch Exception e
            (log/errorf e "failed to checkpoint at shutdown for %s" @shard-id)))))))

(defn shard-record-processor-factory [consume
                                      checkpoint-every-ms]
  (reify ShardRecordProcessorFactory
    (shardRecordProcessor [this]
      (shard-record-processor consume checkpoint-every-ms))))

(defn validate-consumer-args [{:keys [stream-name
                                      application-name
                                      consume] :as args}]
  (when-not (or stream-name
                application-name
                consume)
    (throw (RuntimeException.
             (str "kinesis consumer needs at least these 3 pieces to start: stream name, application name and a consume function. "
                  args))))
  (when-not (fn? consume)
    (throw (RuntimeException.
             (str "'consume' needs to be a _function_ that is capable of taking a batch of kinesis records: "
                  args)))))

(defn echo
  "echo consume function.
   expects records to be UTF-8 decoded"
  [records]
  (let [decode (make-utf8-decoder)]
    (doseq [record records]
      (log/infof "got a record: %s, data: %s\n"
                 record
                 (decode record)))))

(defn show-leases [consumer]
  (-> consumer :scheduler .leaseRefresher .listLeases))

(defn make-lease [lease-key]
  (doto
    (Lease.)
    (.leaseKey lease-key)))

(defn find-leases-for-stream [scheduler
                              stream]
  (let [stream-identifier (if (tracker/multi-stream? scheduler)
                            (tracker/multi-stream-identifier stream)
                            (tracker/single-stream-identifier stream))]
    (-> scheduler
        .leaseRefresher
        (.listLeasesForStream stream-identifier))))

(defn delete-lease [scheduler
                    lease]
  (log/infof "deleting lease: %s for %s application"
             (.leaseKey lease)
             (.applicationName scheduler))
  (-> scheduler
      .leaseRefresher
      (.deleteLease lease)))

(defn delete-leases-for-stream [scheduler
                                stream]
  (log/infof "deleting leases for stream \"%s\", application \"%s\""
             stream
             (.applicationName scheduler))
  (->> stream
       (find-leases-for-stream scheduler)
       (mapv (partial delete-lease scheduler))))

(defn delete-all-leases [scheduler]
  (log/infof "deleting leases for \"%s\" application"
             (.applicationName scheduler))
  (-> scheduler .leaseRefresher .deleteAll))

(defn prep-initial-position
  "if there is an existing lease, by default, kinesis client will ignore an initial position that is provided.
   this deletes existing leases for this app (a.k.a. consumer group) in case there is a position provided"
  [scheduler
   aws-account-number
   {:keys [start-from
           delete-leases?] :as stream}]
  (when-let [{:keys [position]} start-from]
    (when (or delete-leases?
              ;; in order for :trim-horizon and :at-timestamp initial position to take effect
              ;; leases need to be deleted
              (#{:trim-horizon
                 :at-timestamp} position))
      (log/infof "about to delete a leases for %s"
                 stream)
      (delete-leases-for-stream scheduler
                                (assoc stream :aws-account-number
                                              aws-account-number)))))

(defn start-consumer [{:keys [application-name
                              aws-account-number
                              streams
                              creds               ;; in a form of {:access-key-id "foo" :secret-access-key "bar"}
                              region
                              consume
                              checkpoint-every-ms
                              config]             ;; TODO: use this placeholder to exand on the ConfigsBuilder defaults
                       :as opts}]
  (validate-consumer-args opts)
  (let [config-args (merge opts
                           {:kinesis-client (aws/make-kinesis-client opts)
                            :dynamo-client (aws/make-dynamo-db-client opts)
                            :cloud-watch-client (aws/make-cloud-watch-client opts)
                            :record-processor-factory (shard-record-processor-factory consume
                                                                                      checkpoint-every-ms)})
        scheduler (-> (make-config config-args)
                      make-scheduler)]
    (mapv (partial prep-initial-position scheduler
                                         aws-account-number)
          streams)
    {:scheduler scheduler
     :executor (t/start-within-thread scheduler)}))

(defn stop-consumer [{:keys [scheduler
                             executor]}]
  ;; NOTE: current "expected" :( error on graceful shutdown: https://github.com/awslabs/amazon-kinesis-client/issues/914
  ;; (.startGracefulShutdown scheduler)
  (try (.shutdown scheduler)
       (catch Exception e
         (log/warn "could not cleanly shutdown the scheduler for"
                   (.applicationName scheduler)
                   "application due to" e)))
  (try (.shutdown executor)
       (catch Exception e
         (log/warn "could not cleanly shutdown the executor for"
                   (.applicationName scheduler)
                   "application due to" e))))

File: dev/java/kinetic/src/main/java/com/tolitius/Consumer.java
package com.tolitius;

/**
 * building on the aws official example:
 *          https://docs.aws.amazon.com/streams/latest/dev/kcl2-standard-consumer-java-example.html
 */

import org.apache.commons.lang3.ObjectUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.slf4j.MDC;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient;
import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient;
import software.amazon.awssdk.services.kinesis.KinesisAsyncClient;
import software.amazon.kinesis.common.ConfigsBuilder;
import software.amazon.kinesis.common.InitialPositionInStreamExtended;
import software.amazon.kinesis.common.KinesisClientUtil;
import software.amazon.kinesis.common.StreamIdentifier;
import software.amazon.kinesis.coordinator.Scheduler;
import software.amazon.kinesis.exceptions.InvalidStateException;
import software.amazon.kinesis.exceptions.ShutdownException;
import software.amazon.kinesis.leases.exceptions.DependencyException;
import software.amazon.kinesis.leases.exceptions.ProvisionedThroughputException;
import software.amazon.kinesis.lifecycle.events.*;
import software.amazon.kinesis.processor.ShardRecordProcessor;
import software.amazon.kinesis.processor.ShardRecordProcessorFactory;
import software.amazon.kinesis.processor.SingleStreamTracker;
import software.amazon.kinesis.retrieval.KinesisClientRecord;
import software.amazon.kinesis.retrieval.polling.PollingConfig;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Date;
import java.util.UUID;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static software.amazon.kinesis.common.InitialPositionInStream.TRIM_HORIZON;

/**
 * This class will run a simple app that uses the KCL to read data and uses the AWS SDK to publish data.
 * Before running this program you must first create a Kinesis stream through the AWS console or AWS SDK.
 */
public class Consumer {

    private static final Logger log = LoggerFactory.getLogger(Consumer.class);

    /**
     * Invoke the main method with 2 args: the stream name and (optionally) the region.
     * Verifies valid inputs and then starts running the app.
     */
    public static void main(String... args) {
        if (args.length < 1) {
            log.error("At a minimum, the stream name is required as the first argument. The Region may be specified as the second argument.");
            System.exit(1);
        }

        String applicationName = args[0];
        String streamName = args[1];
        String region = null;
        if (args.length > 2) {
            region = args[2];
        }

        new Consumer(applicationName, streamName, region).run();
    }

    private final String applicationName;
    private final String streamName;
    private final Region region;
    private final KinesisAsyncClient kinesisClient;

    /**
     * Constructor sets streamName and region. It also creates a KinesisClient object to send data to Kinesis.
     * This KinesisClient is used to send dummy data so that the consumer has something to read; it is also used
     * indirectly by the KCL to handle the consumption of the data.
     */
    private Consumer(String applicationName, String streamName, String region) {
        this.applicationName = applicationName;
        this.streamName = streamName;
        this.region = Region.of(ObjectUtils.firstNonNull(region, "us-east-1"));
        this.kinesisClient = KinesisClientUtil.createKinesisAsyncClient(KinesisAsyncClient.builder().region(this.region));
    }

    private void run() {

        /**
         * Sets up configuration for the KCL, including DynamoDB and CloudWatch dependencies. The final argument, a
         * ShardRecordProcessorFactory, is where the logic for record processing lives, and is located in a private
         * class below.
         */
        DynamoDbAsyncClient dynamoClient = DynamoDbAsyncClient.builder().region(region).build();
        CloudWatchAsyncClient cloudWatchClient = CloudWatchAsyncClient.builder().region(region).build();

        var daysAgo = Date.from(Instant.now().minus(1, ChronoUnit.DAYS));

        var startFrom = InitialPositionInStreamExtended.newInitialPosition(TRIM_HORIZON);
//        var startFrom = InitialPositionInStreamExtended.newInitialPositionAtTimestamp(daysAgo);

        var streamTracker =
                new SingleStreamTracker(StreamIdentifier.singleStreamInstance(streamName),
                startFrom);
        
        ConfigsBuilder configsBuilder =
                new ConfigsBuilder(streamTracker,
                                   applicationName,
                                   kinesisClient,
                                   dynamoClient,
                                   cloudWatchClient,
                                   UUID.randomUUID().toString(),
                                   new SampleRecordProcessorFactory());

        /**
         * The Scheduler (also called Worker in earlier versions of the KCL) is the entry point to the KCL. This
         * instance is configured with defaults provided by the ConfigsBuilder.
         */
        Scheduler scheduler = new Scheduler(
                configsBuilder.checkpointConfig(),
                configsBuilder.coordinatorConfig(),
                configsBuilder.leaseManagementConfig(),
                configsBuilder.lifecycleConfig(),
                configsBuilder.metricsConfig(),
                configsBuilder.processorConfig(),
                configsBuilder.retrievalConfig().retrievalSpecificConfig(new PollingConfig(streamName, kinesisClient))
        );

        var leaseRefresher  = scheduler.leaseRefresher();

        try {
            leaseRefresher.deleteAll();
        } catch (DependencyException e) {
            e.printStackTrace();
        } catch (software.amazon.kinesis.leases.exceptions.InvalidStateException e) {
            e.printStackTrace();
        } catch (ProvisionedThroughputException e) {
            e.printStackTrace();
        }

        /**
         * Kickoff the Scheduler. Record processing of the stream of dummy data will continue indefinitely
         * until an exit is triggered.
         */
        Thread schedulerThread = new Thread(scheduler);
        schedulerThread.setDaemon(true);
        schedulerThread.start();

        /**
         * Allows termination of app by pressing Enter.
         */
        System.out.println("Press enter to shutdown");
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        try {
            reader.readLine();
        } catch (IOException ioex) {
            log.error("Caught exception while waiting for confirm. Shutting down.", ioex);
        }

        /**
         * Stops consuming data. Finishes processing the current batch of data already received from Kinesis
         * before shutting down.
         */
        Future<Boolean> gracefulShutdownFuture = scheduler.startGracefulShutdown();
        log.info("Waiting up to 20 seconds for shutdown to complete.");
        try {
            gracefulShutdownFuture.get(20, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            log.info("Interrupted while waiting for graceful shutdown. Continuing.");
        } catch (ExecutionException e) {
            log.error("Exception while executing graceful shutdown.", e);
        } catch (TimeoutException e) {
            log.error("Timeout while waiting for shutdown.  Scheduler may not have exited.");
        }
        log.info("Completed, shutting down now.");
    }

    private static class SampleRecordProcessorFactory implements ShardRecordProcessorFactory {
        public ShardRecordProcessor shardRecordProcessor() {
            return new SampleRecordProcessor();
        }
    }

    /**
     * The implementation of the ShardRecordProcessor interface is where the heart of the record processing logic lives.
     * In this example all we do to 'process' is log info about the records.
     */
    private static class SampleRecordProcessor implements ShardRecordProcessor {

        private static final String SHARD_ID_MDC_KEY = "ShardId";

        private static final Logger log = LoggerFactory.getLogger(SampleRecordProcessor.class);

        private String shardId;

        /**
         * Invoked by the KCL before data records are delivered to the ShardRecordProcessor instance (via
         * processRecords). In this example we do nothing except some logging.
         *
         * @param initializationInput Provides information related to initialization.
         */
        public void initialize(InitializationInput initializationInput) {
            shardId = initializationInput.shardId();
            MDC.put(SHARD_ID_MDC_KEY, shardId);
            try {
                log.info("Initializing @ Sequence: {}", initializationInput.extendedSequenceNumber());
            } finally {
                MDC.remove(SHARD_ID_MDC_KEY);
            }
        }

        public String readRecordData (KinesisClientRecord record) {

            ByteBuffer data = record.data();
            final byte[] bytes = new byte[data.remaining()];
            data.duplicate().get(bytes);

            return new String(bytes);
        }

        /**
         * Handles record processing logic. The Amazon Kinesis Client Library will invoke this method to deliver
         * data records to the application. In this example we simply log our records.
         *
         * @param processRecordsInput Provides the records to be processed as well as information and capabilities
         *                            related to them (e.g. checkpointing).
         */
        public void processRecords(ProcessRecordsInput processRecordsInput) {
            MDC.put(SHARD_ID_MDC_KEY, shardId);
            try {
                log.info("Processing {} record(s)", processRecordsInput.records().size());
                processRecordsInput.records().forEach(
                        r -> log.info("\n\nProcessing record pk: {} -- Seq: {}, --data {}",
                                r.partitionKey(),
                                r.sequenceNumber(),
                                readRecordData(r)));
            } catch (Throwable t) {
                log.error("Caught throwable while processing records. Aborting.");
                Runtime.getRuntime().halt(1);
            } finally {
                MDC.remove(SHARD_ID_MDC_KEY);
            }
        }

        /** Called when the lease tied to this record processor has been lost. Once the lease has been lost,
         * the record processor can no longer checkpoint.
         *
         * @param leaseLostInput Provides access to functions and data related to the loss of the lease.
         */
        public void leaseLost(LeaseLostInput leaseLostInput) {
            MDC.put(SHARD_ID_MDC_KEY, shardId);
            try {
                log.info("Lost lease, so terminating.");
            } finally {
                MDC.remove(SHARD_ID_MDC_KEY);
            }
        }

        /**
         * Called when all data on this shard has been processed. Checkpointing must occur in the method for record
         * processing to be considered complete; an exception will be thrown otherwise.
         *
         * @param shardEndedInput Provides access to a checkpointer method for completing processing of the shard.
         */
        public void shardEnded(ShardEndedInput shardEndedInput) {
            MDC.put(SHARD_ID_MDC_KEY, shardId);
            try {
                log.info("Reached shard end checkpointing.");
                shardEndedInput.checkpointer().checkpoint();
            } catch (ShutdownException | InvalidStateException e) {
                log.error("Exception while checkpointing at shard end. Giving up.", e);
            } finally {
                MDC.remove(SHARD_ID_MDC_KEY);
            }
        }

        /**
         * Invoked when Scheduler has been requested to shut down (i.e. we decide to stop running the app by pressing
         * Enter). Checkpoints and logs the data a final time.
         *
         * @param shutdownRequestedInput Provides access to a checkpointer, allowing a record processor to checkpoint
         *                               before the shutdown is completed.
         */
        public void shutdownRequested(ShutdownRequestedInput shutdownRequestedInput) {
            MDC.put(SHARD_ID_MDC_KEY, shardId);
            try {
                log.info("Scheduler is shutting down, checkpointing.");
                shutdownRequestedInput.checkpointer().checkpoint();
            } catch (ShutdownException | InvalidStateException e) {
                log.error("Exception while checkpointing at requested shutdown. Giving up.", e);
            } finally {
                MDC.remove(SHARD_ID_MDC_KEY);
            }
        }
    }

}
